File: source/includes/fts/queries/complex-fts-query.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;
using MongoDB.Bson.Serialization.Conventions;
using MongoDB.Driver;
using MongoDB.Driver.Search;

public class ComplexQuery
{
    private const string MongoConnectionString = "<connection-string>";

    public static void Main(string[] args)
    {
        // allow automapping of the camelCase database fields to our MovieDocument
        var camelCaseConvention = new ConventionPack { new CamelCaseElementNameConvention() };
        ConventionRegistry.Register("CamelCase", camelCaseConvention, type => true);

        // connect to your Atlas cluster
        var mongoClient = new MongoClient(MongoConnectionString);
        var mflixDatabase = mongoClient.GetDatabase("sample_mflix");
        var moviesCollection = mflixDatabase.GetCollection<MovieDocument>("movies");

        // declare data for the compound query
        string[] terms = { "Hawaii", "Alaska" };
        string[] genres = { "Comedy", "Romance" };
        string[] titles = { "Beach", "Snow" };

        // define and run pipeline
        var results = moviesCollection.Aggregate()
            .Search(Builders<MovieDocument>.Search.Compound()
                .Must(Builders<MovieDocument>.Search.Text(movie => movie.Plot, terms))
                .Must(Builders<MovieDocument>.Search.Regex(movie => movie.Plot, "([0-9]{4})", true))
                .MustNot((Builders<MovieDocument>.Search.Text(movie => movie.Genres, genres)))
                .MustNot((Builders<MovieDocument>.Search.Text(movie => movie.Title, titles))))
            .Project<MovieDocument>(Builders<MovieDocument>.Projection
                .Include(movie => movie.Genres)
                .Include(movie => movie.Plot)
                .Include(movie => movie.Title)
                .Exclude(movie => movie.Id))
            .ToList();

        // print results
        foreach (var movie in results)
        {
            Console.WriteLine(movie.ToJson());
        }
    }
}

[BsonIgnoreExtraElements]
public class MovieDocument
{
    [BsonIgnoreIfDefault]
    public ObjectId Id { get; set; }
    public string[] Genres { get; set; }
    public string Plot { get; set; }
    public string Title { get; set; }
}

File: source/includes/fts/queries/complex-fts-query.go
package main
import (
	"context"
	"fmt"
	"time"

	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)
// define structure of movies collection
type MovieCollection struct {
  title   string      `bson:"Title,omitempty"`
  plot    []string    `bson:"Plot,omitempty"`
}
func main() {
  var err error
  // connect to the Atlas cluster
  ctx := context.Background()
  client, err := mongo.Connect(ctx, options.Client().ApplyURI("<connection-string>"))
  if err != nil {
	panic(err)
  }
  defer client.Disconnect(ctx)
  // set namespace
  collection := client.Database("sample_mflix").Collection("movies")
  // define pipeline
  searchStage := bson.D{{"$search", bson.M{
	"compound": bson.M{
	  "must": bson.A{
		bson.M{
		  "text": bson.D{
			{"path", "plot"},{"query", []string{"Hawaii", "Alaska"}},
		  },
		},
		bson.M{
		  "regex": bson.M{
		  "path": "plot","query": "([0-9]{4})","allowAnalyzedField": true,
		  },
		},
	  },
	  "mustNot": bson.A{
		bson.D{
		  {"text", bson.M{
			 "path": "genres","query": []string{"Comedy", "Romance"},
		  }},
		  {"text", bson.M{
			"path": "title","query": []string{"Beach", "Snow"},
		  }},
		},
	  },
	},
  }}}
  projectStage := bson.D{{"$project", bson.D{{"title", 1}, {"plot", 1}, {"genres", 1}, {"_id", 0}}}}
  // specify the amount of time the operation can run on the server
  opts := options.Aggregate().SetMaxTime(5 * time.Second)
  // run pipeline
  cursor, err := collection.Aggregate(ctx, mongo.Pipeline{searchStage, projectStage}, opts)
  if err != nil {
	panic(err)
  }
  // print results
  var results []bson.D
  if err = cursor.All(context.TODO(), &results); err != nil {
	panic(err)
  }
  for _, result := range results {
	fmt.Println(result)
  }
}
